src/tests/JIT/Performance/JittedMethodsCountingTest/JittedMethodsCountingTest.cs

```diff
@@ -0,0 +1,129 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using Xunit;
+
+public class JittedMethodsCountingTest
+{
+    private const int MAX_JITTED_METHODS_ACCEPTED = 120;
```

Comment:
120 seems like a pretty high number to allow -- do we want to set this lower?

Reply:
Empty app on Linux/macOS x64/arm64 is expected to produce just [1 jit compilation](https://github.com/dotnet/runtime/pull/90412#issuecomment-1679822542) (for Main) 🙂 
On windows there can be more due to background ETW sessions, but not too much (like 16 in total or so)

For Console.WriteLine it's tricky because `System.Console` is not prejitted in dotnet/runtime, hence, all methods in it are always jitted

Reply:
I must admit I'm not too familiarized with the expected jitted methods counts. What would be a good expected range for a HelloWorld app like the one added in this project?

Reply:
I think we just need a baseline and go from there. I recall that in my experiments with startup of the "dotnet new webapi" I observed about 50 methods being jitted at runtime when compiled as composite with the MIBC optimization data i.o.w. in more or less "the best possible configuration".

Reply:
(Without the MIBC data it was almost 200 IIRC.)

Reply:
This test is only protecting against something going completely wrong with crossgen of System.Private.CoreLib. It does not run on actual shipping bits as Egor mentioned and it runs in unstable environment (background ETW sessions) that needs to allow for margin of error.

Is this test going to catch the two issues that sneaked through that motivated us to add it?

Reply:
> I think we just need a baseline and go from there.

The construction of this test and where we have added it does not really allow us to go from here.

Reply:
Hmm, I don't understand, can you please clarify? I would assume that, once this test has been merged in, anyone is welcome to establish a tighter upper bound for the number of jitted methods provided their lab testing passes. If you're alluding to the fact that the test is supposed to run in various configurations (with tiered compilation on vs. off, on different OSes and architectures), I guess it's up to further discussion if we're fine with something like a "greatest common divisor" in the sense of maximum methods jitted in all configurations or if we need something like a table with the expected method counts per various build platforms / execution modes.

Reply:
FWIW I just tried with "dotnet new console" compiled using ".NET 8.0 preview 7 SDK" in release mode (Windows x64) with ```PublishReadyToRun=true``` and ```PublishReadyToRunComposite=true``` (without the MIBC optimization data) and I'm seeing 109 methods JITted at runtime.

Reply:
If we come to the conclusion that ```Console.WriteLine``` is skewing the number of jitted methods in an undesirable manner, I guess it should be no big deal for Ivan to delete the one line from his test app :-); in a way it may be an even better approximation of a really empty app doing just the runtime initialization and shutdown.

Reply:
> I don't understand, can you please clarify?

We need to be clear about the purpose of this test.

As it is setup, this test protects against something going very wrong with crossgen of System.Private.CoreLib. Is this the intended long-term purpose of this test? If not, what is it instead?

I do not think we should be treating this test as a perf test. It does not test the actual product bits and it would not scale well to larger apps like `dotnet webapi`. Also, we should not be optimizing number of JITed methods without looking at the wall clock time that these methods took to JIT. The primary metric that we care about is the startup time and the time that the methods participating in startup took to JIT. The number of the JITed methods is a secondary metric and it can be easily skewed (e.g. merging two small methods into one big method). There are classic anecdotes in MS perf community about how focus on secondary metric resulted into the primary metric regressing, we do not want to the next one.

I think we should follow the same approach for startup as we do with other benchmarks and allow the perflab to be the primary place that it is measured and tracked. I would like to see the nice graphs that show how the time spend JITing methods and the number of JITed methods in `dotnet new console` or `dotnet new webapi` startup kept changing over the course of the release. Does something like that exist already?

Note that the existing tests under `src\tests\JIT\Performance` where this test is being added are not run as a performance benchmarks. These tests exist in the test tree to ensure they are working and to allow JIT team infra see the asm diffs for them. The equivalent benchmark versions of these tests live in perf benchmarks repo under https://github.com/dotnet/performance/tree/main/src/benchmarks/micro/runtime and these copies run in the perflab.

Reply:
> We want to monitor the number of jitted methods. Nothing more, nothing less. 

This is not clear explanation for me. It explains the work. It does not explain what we want to achieve.

> In that case, where would you suggest placing it?

If you want to treat it as a perf test, it should live in https://github.com/dotnet/performance along with all other runtime perf benchmarks that we have.

Reply:
> We can always add new apps and make it more configurable, which is the longer term goal. 

The core runtime tests are not the best place for adding end-to-end (perf) tests.

Reply:
I understand your point, even if I do not agree with it. But I have a question. Why didn't you provide these ideas when you gave your first feedback? I find it difficult to digest to get approved only for at the end be told that basically it should be done an entirely different way. Even in an entirely different place.

Reply:
From my perspective the intent of this test in the runtime repo is to smoke test to ensure that R2R of SPCL is not completely broken. Currently we would only notice such a break when microbenchmarks or TechEmpower flag a large startup regression. My thinking would be that we are far more lenient with the allowed number of JITted methods.

Reply:
as mentioned here: https://github.com/dotnet/runtime/issues/88533#issuecomment-1626319472.

Reply:
We already have simliar smoke-tests, right? like that NAOT test where we check that an empty app's size is not terribly big in dotnet/runtime repo - is it real-world? No, but better than nothing and, afair, it already helped once. I agree that this test as is won't help us to find cases when something goes wrong but not terribly (e.g. 100 methods regressed). For that we still need to apply efforts elsewhere: I'd want to have a separate graph for "methods jitted" counter at https://aka.ms/aspnet/benchmarks so once startup regresses again I will be able at least quickly validate that it's not because of R2R. A visualization on dotnet/performance would work too.

Reply:
> As for where the runtime repo version should be placed, considering what you explained about `src/tests/JIT/Performance`, then I agree it's not the best place. In that case, where would you suggest placing it?

I agree that "JIT/Performance" is not the right place.

How about src\tests\readytorun\...?

Reply:
> Why didn't you provide these ideas when you gave your first feedback? 

I have provided similar feedback in the issue discussion - look for my comments in https://github.com/dotnet/runtime/issues/88533.

> the test is ok to have IMO

Agree - this test as a smoke test that will catch things going terribly wrong. It is not a perf gate.

Reply:
> We already have similar smoke-tests in dotnet/runtime, right? like that NAOT test where we check that an empty app's size is not terribly big in dotnet/runtime repo - is it real-world? No, but better than nothing and, afair, it already helped once. I agree that this test as is won't help us to find cases when something goes wrong but not terribly (e.g. 100 methods regressed). For that we still need to apply efforts elsewhere: I'd want to have a separate graph for "methods jitted" counter at https://aka.ms/aspnet/benchmarks so once startup regresses again I will be able at least quickly validate that it's not because of some obvious R2R issue. A visualization on dotnet/performance would work too.
> 
> tldr: the test is ok to have IMO, additional tests in dotnet/performance and PerfLab would be nice to have too (where we work with SDK)

Thanks for your feedback @EgorBo. I agree having a visual graph would certainly help us catch regressions-in-the-making sooner. Then I have a follow up question for you and @AndyAyersMS. What would be a good number of jitted methods that if exceeded, we can confirm something went terribly wrong and therefore fail this test?

As for the dashboard and graphs, I will reach out to Sebastien to get it to run periodically in TE alongside the other perf tests we already run (e.g. ASP.NET Composite Containers).

> How about src\tests\readytorun...?

I think this is a good place to put this one @BruceForstall. I'll move it and rerun the lab tests.

Reply:
> What would be a good number of jitted methods that if exceeded, we can confirm something went terribly wrong and therefore fail this test?

It should be the lowest number that makes the test pass reliably with some margin. What is the highest number of methods that you see being JITed across different configurations? Add 20 to it and use that as the limit.

> I will reach out to Sebastien to get it to run periodically in TE alongside the other perf tests we already run (e.g. ASP.NET Composite Containers)

A trivial `Console.WriteLine` startup perf test has nothing to do with ASP.NET. It does not sound right to add it to ASP.NET perf infra for TE.

We have startup-oriented scenario perf measurements under https://github.com/dotnet/performance/tree/main/src/tools/ScenarioMeasurement. We should have a guidance for when to add a scenario measurement to this location vs. to the TE infrastructure. Please check with Sebastien and Scott Blomquist about that.